File: benchmarks/_foreign/mandel.py
import time
import math

def norm2(x_re, x_im):
    return x_re * x_re - x_im * (-x_im)

def im_abs(x_re, x_im):
    return math.sqrt(norm2(x_re, x_im))

def level(x, y):
    c_re = x
    c_im = y
    l = 0
    z_re = c_re 
    z_im = c_im
    
    while True:
        t_re = z_re * z_re - z_im * z_im
        t_im = z_re * z_im + z_im * z_re
        
        z_re = t_re + c_re
        z_im = t_im + c_im 

        l = l + 1

        if im_abs(z_re, z_im) > 2 or l > 255:
            return l - 1

for outer in range(15):
    start = time.time()

    xmin = -2.0
    xmax = 2.0
    ymin = -2.0
    ymax = 2.0
    n = 256 

    dx = (xmax - xmin) / n
    dy = (ymax - ymin) / n

    s = 0
    for i in range(n):
        x = xmin + i * dx
        for j in range(n):
            y = ymin + j * dy
            s = s + level(x, y)

    print("Time elapsed is", (time.time() - start) * 1000, "ms |", s)

File: benchmarks/_foreign/mandel.js
function norm2(re, im) {
    return re * re - im * (-im)
}

function xabs(re, im) {
    return Math.sqrt(norm2(re, im))
}

function level(x, y) {
    let cre = x
    let cim = y
    let zre = cre
    let zim = cim

    for (let i = 0; i < 255; i++) {
        let tre = zre * zre - zim * zim
        let tim = zre * zim + zim * zre

        zre = tre + cre
        zim = tim + cim

        if(xabs(zre, zim) > 2) {
            return i
        }
    }

    return 255
}

for (let outer = 0; outer < 15; outer++) {
    var start_time = Date.now()

    let xmin = -2
    let xmax = 2
    let ymin = -2
    let ymax = 2
    let n = 256

    let dx = 4 / n
    let dy = 4 / n

    let result = 0
    for(let xi = 0; xi < n; xi++) {
        let x = xmin + xi * dx
        for(let yi = 0; yi < n; yi++) {
            let y = ymin + yi * dy
            result += level(x, y)
        }
    }

    var end_time = Date.now()
    var duration_ms = end_time - start_time
    console.log("10m range() iterations took", duration_ms, "ms", "|", result)
}

